controlplane/src/gatewayclass_controller.rs

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use futures::StreamExt;
+use std::{
+    ops::Sub,
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+use crate::*;
+use gateway_api::apis::standard::gatewayclasses::GatewayClass;
+use kube::{
+    api::{Api, ListParams},
+    runtime::{controller::Action, watcher::Config, Controller},
+};
+
+use gatewayclass_utils::*;
+use tracing::*;
+
+pub async fn reconcile(gateway_class: Arc<GatewayClass>, ctx: Arc<Context>) -> Result<Action> {
+    let start = Instant::now();
+    let client = ctx.client.clone();
+    let name = gateway_class
+        .metadata
+        .name
+        .clone()
+        .ok_or(Error::InvalidConfigError("invalid name".to_string()))?;
+
+    let gatewayclass_api = Api::<GatewayClass>::all(client);
+    let mut gwc = GatewayClass {
+        metadata: gateway_class.metadata.clone(),
+        spec: gateway_class.spec.clone(),
+        status: gateway_class.status.clone(),
+        // NOTE: Am I missing anything else here?
+    };
```

Comment:
If the goal is to make a full clone, you should be able to do that with [Deref](https://doc.rust-lang.org/std/sync/struct.Arc.html#deref-behavior) and clone:
```suggestion
    let mut gwc = (*gateway_class).clone();
```

Reply:
This is indeed the aim, unless I can get away with not cloning at all. But we ultimately would need a thread-safe mutable reference (`Arc<Mutex<GatewayClass>>` to our gateway_class input then, but AFAICT from the kube.rs docs, `reconcile` will always be passed an `Arc<>` of the resource it's watching. Is that so?

Reply:
It will always be an Arc, so if you want the inner of an arc you can do a deref clone (as shown above) OR an `as_ref().clone()`.